File: python/331.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        st = []
        for node in preorder.split(","):
            st.append(node)
            while len(st) > 2 and st[-1] == st[-2] == '#' and st[-3] != '#':
                st.pop()
                st.pop()
                st.pop()
                st.append('#')
        return len(st) == 1 and st[0] == '#'

File: go/331.go
package main

import (
	"strings"
)

func isValidSerialization(preorder string) bool {
	st := make([]string, 0)
	nodes := strings.Split(preorder, ",")
	for _, node := range nodes {
		st = append(st, node)
		for len(st) > 2 && st[len(st)-1] == "#" && st[len(st)-2] == "#" && st[len(st)-3] != "#" {
			st = st[:len(st)-3]
			st = append(st, "#")
		}
	}
	return len(st) == 1 && st[0] == "#"
}
